### PR TITLE
Enable logging

### LINKED ISSUE #340: [FEA] Enable logging
Logging via spdlog was added in #297 , but it is not enabled yet. Enable it via memory_manager.cpp so it can be enabled using the current Python API.

### PR DESCRIPTION
**Pull request by harrism**
Fixes #340 

Plumbs `logging_resource_adaptor` in `memory_manager.cpp` so that it can be enabled by setting `logging = True` in Python `rmm.reinitialize()`.

Example

```
import os

def set_rmm_log_file():
    os.environ["RMM_LOG_FILE"] = "test.txt"
    return os.environ["RMM_LOG_FILE"]

set_rmm_log_file()

import rmm

rmm.reinitialize(
    pool_allocator=True,
    managed_memory=False,
    initial_pool_size=2**31,
    logging=True
    )

rmm.DeviceBuffer(size=1048576)

print(os.environ.get("RMM_LOG_FILE"))
```

Resulting log file:

```
Time,Action,Pointer,Size,Stream
22:37:25:738700,allocate,0x7f68a2000000,1048576,0
22:37:25:738725,free,0x7f68a2000000,1048576,0
```

### CODE REVIEW

File: src/memory_manager.cpp [hunk 1]
```diff
@@ -102,20 +123,21 @@ void Manager::initialize(const rmmOptions_t* new_options) {
...
+  bool enable_logging = getOptions().enable_logging;
+
   if (usePoolAllocator()) {
     auto pool_size = getOptions().initial_pool_size;
     auto const& devices = getOptions().devices;
+
     if (useManagedMemory()) {
-      initialized_resource.reset(
-          new rmm::mr::cnmem_managed_memory_resource(pool_size, devices));
+      reset_resource(initialized_resource, new pool_managed_mr(pool_size, devices), enable_logging);
```
Comment (jrhemstad) [inline src/memory_manager.cpp:133]: When logging is enabled, the upstream resource won't get cleaned up when the `unique_ptr` is destroyed, it will only clean up the logger, and the logger doesn't do anything to cleanup the upstream. 

To solve this, we can either use a custom deleter in `reset_resource` that destroys both the upstream and the logger, or we can add another `unique_ptr<mr>` alongside `initialized_resource`. So something like this (but could use some streamlining):
```
   initialized_resource.reset(new cuda_mr{});
   if(enabled_logging){
      logging_adaptor.reset(new logging_resource_adaptor<cuda_memory_resource>(mr);
      set_default_resource(logging_adaptor.get());
   }
```

Comment (harrism) [reply src/memory_manager.cpp:133]: It's funny I thought of this memory leak during your talk today too...

Comment (harrism) [reply src/memory_manager.cpp:133]: Also just realized you already had a "do not merge" PR open for this!

Comment (jrhemstad) [reply src/memory_manager.cpp:133]: My solution was a hackier version of what you're doing. My PR _always_ enables logging whether the user requested it or not.

**Comment by jakirkham** [comment]
Thanks Mark! This is great 😄

cc @kkraus14 @quasiben (for awareness)

File: src/memory_manager.cpp (hunk 1 above)
Comment (harrism) [reply src/memory_manager.cpp]: Hey @jrhemstad I fixed this in my latest commit, but I'm not happy with it. I had to go out of my way to create the `reset_resource` template function just because I wanted to avoid a vtable lookup for every call to the logging_adaptor. This was made tricky because of the fact that I couldn't get memory_manager.hpp to compile if it includes `logging_resource_adaptor.hpp` due to the spdlog dependency (moreover, it's a template class, so it would be tricky to have a member pointer  to it anyway).

But the end result is I think the type is lost when it's stored into the unique_ptr member anyway, which means a user of get_default_resource will incur not one but two vtable lookups per call to allocate...

Comment (jrhemstad) [reply src/memory_manager.cpp]: Looking at the way you've implemented it, there will only be one virtual function call in `get_default_resource()->allocate(n)`. 

This will incur a vtable lookup to call `logging_resource_adaptor::allocate(n)`, but then when `logging_resource_adaptor::allocate(n)` calls `upstream.allocate(n)`, the upstream call will not incur a vtable lookup since the logging adaptor knows the concrete type of `upstream`. 

There may be a cleaner way to accomplish the same thing, but I think what you've got is fine given that it is intended to be temporary.

Comment (harrism) [reply src/memory_manager.cpp]: Yes, good point. I was looking at the `logging_adaptor` member, since it is abstract.

**Comment by abellina** [comment]
@jrhemstad @harrism in the case where `RMM_LOG_FILE` is undefined, I am seeing failures while initializing RMM (`rmmInitialize` with `enable_logging=true`).

```
basic_string::_M_construct null not valid
```

This (I believe) is because it is trying to construct a string from `std::getenv`, and getting null:

```
std::string const& filename = std::getenv("RMM_LOG_FILE")
```

It would be nice NOT to generate a file via this, unless you have to as well. Unless `getLog` is going away?

**Comment by abellina** [comment]
@jlowe posted https://github.com/rapidsai/rmm/issues/344 to handle https://github.com/rapidsai/rmm/pull/341#issuecomment-611318572

**Comment by harrism** [comment]
@abellina sorry about that. Please file an issue.

File: src/memory_manager.cpp [hunk 2]
```diff
@@ -20,8 +20,17 @@
 #include <rmm/mr/device/cuda_memory_resource.hpp>
 #include <rmm/mr/device/default_memory_resource.hpp>
 #include <rmm/mr/device/managed_memory_resource.hpp>
+#include <rmm/mr/device/logging_resource_adaptor.hpp>
 
 namespace rmm {
+
+using cuda_mr = rmm::mr::cuda_memory_resource;
+using pool_mr = rmm::mr::cnmem_memory_resource;
+using managed_mr = rmm::mr::cnmem_managed_memory_resource;
```
Comment (cwharris) [inline src/memory_manager.cpp:29]: Was this change intentional, or should we still be using using `rmm::mr::managed_memory_resource` ?
